File: julia/Day5.jl
module Day5

export Submarine, on_input!, output

const FIELD_SIZE = 1000

mutable struct Submarine
    vents_field_no_diagonal::Matrix{Int32}
    vents_field::Matrix{Int32}

    Submarine() = new(zeros(Int32, FIELD_SIZE, FIELD_SIZE), zeros(Int32, FIELD_SIZE, FIELD_SIZE))
end

function on_input!(s::Submarine, a::AbstractString)
    x0, y0, x1, y1 = Iterators.flatten(split(a, " -> ") .|> (x) -> split(x, ",")) .|> (x) -> 1 + parse(Int32, x)

    mark_vents!(s, x0, y0, x1, y1)
end

function mark_vents!(s::Submarine, x0::Integer, y0::Integer, x1::Integer, y1::Integer)
    xrange = build_range(x0, x1)
    yrange = build_range(y0, y1)

    # If x0 == x1 or y0 == y1, we can cycle forever the range that will only yield x0 or y0 to be able to zip() both ranges.
    if x0 == x1
        xrange = Iterators.cycle(xrange)
    end

    if y0 == y1
        yrange = Iterators.cycle(yrange)
    end

    is_diagonal = x0 != x1 && y0 != y1

    for (x, y) in Iterators.zip(xrange, yrange)
        s.vents_field[x, y] += 1

        if !is_diagonal
            s.vents_field_no_diagonal[x, y] += 1
        end
    end
end

function build_range(a::Integer, b::Integer)
    if a > b
        a:-1:b
    else
        a:b
    end
end

function output(s::Submarine)
    dangerous_areas_no_diagonal = length(filter((x) -> x > 1, s.vents_field_no_diagonal))
    dangerous_areas = length(filter((x) -> x > 1, s.vents_field))

    println("Part 1: $(dangerous_areas_no_diagonal)")
    println("Part 2: $(dangerous_areas)")
end

end

File: rust/src/day5.rs
const FIELD_SIZE: usize = 1000;

pub struct Submarine {
    vents_field: Vec<i32>,
}

impl Submarine {
    pub fn new() -> Self {
        Self {
            vents_field: vec![0; FIELD_SIZE * FIELD_SIZE],
        }
    }

    pub fn on_input(&mut self, result: &str) {
        let vertices: Vec<i32> = result.split(" -> ").map(|x| x.split(",")).flatten().map(|x| x.parse().unwrap()).collect();
        let (x0, y0, x1, y1) = (vertices[0], vertices[1], vertices[2], vertices[3]);

        self.mark_vent(x0, y0, x1, y1);
    }

    fn mark_vent(&mut self, x0: i32, y0: i32, x1: i32, y1: i32) {
        // In case x0 != x1 and y0 != y1 we'll ignore, since part 1 doesn't care about those.

        if x0 == x1 {
            let range = if y0 > y1 {
                y1..(y0 + 1)
            } else {
                y0..(y1 + 1)
            };

            for y in range {
                self.mark_location(x0, y);
            }
        } else if y0 == y1 {
            let range = if x0 > x1 {
                x1..(x0 + 1)
            } else {
                x0..(x1 + 1)
            };

            for x in range {
                self.mark_location(x, y0);
            }
        }
    }

    fn mark_location(&mut self, x: i32, y: i32) {
        let index: usize = (x + y * FIELD_SIZE as i32).try_into().unwrap();

        self.vents_field[index] += 1;
    }

    pub fn output(&self) {
        let dangerous_areas = self.vents_field.iter().filter(|&&x| x > 1).count();

        println!("Part 1: {}", dangerous_areas);
    }
}